enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,504 @@
+- Accidental Eviction of Critical Workloads: The application of NoExecute taints and subsequent enforcement might lead to the accidental eviction of workloads that are critical but were not correctly configured with the necessary tolerations.
+  - Mitigation: Prior to applying NoExecute taints, perform a comprehensive review of all workloads running on the nodes to be protected. Ensure that all essential services and components include the correct tolerations. Utilize dry-run or audit modes of the Validating Admission Policy, if available, to assess the impact before advised enforcement.
+
+### Drawbacks
+
+- Increased Complexity for Cluster Administrators: The introduction of NoExecute taints and the requirement to manage Validating Admission Policies and Bindings may increase the complexity of cluster administration.
+Administrators now need a deeper understanding of how taints, tolerations, and admission policies interact to enforce these constraints.
+
+- Potential for Misconfiguration: The reliance on RBAC rules to exempt certain workloads from scheduling restrictions introduces a risk of misconfiguration, either by applying incorrect rules or failing to update rules as policies evolve.
+
+- Risk of Disruption to Existing Workloads: Applying NoExecute taints to nodes could lead to the eviction of existing workloads that do not have the necessary tolerations, potentially disrupting services.
+
+- Limitation on Feedback Detail from Admission Policies: The feedback provided to users when their pods are rejected due to policy violations may not always be detailed or user-friendly, potentially leading to confusion and delays in troubleshooting.
+
+- Challenges with Third-Party Workloads: Ensuring that third-party operators or helm charts comply with the new scheduling restrictions could be challenging, especially if those workloads require updates to include the necessary tolerations.
+
+## Test Plan
+
+The workflow was manually tested as described in the `Workflow Steps` section. Since these will be the manual steps customers/cluster admins would have to follow, it is validated.
+
+For this to work out we need to be sure that all workloads running in control-plane nodes have the relevant tolerations, so a new e2e tests enforcing that will be added to our test suites.
+
+## Graduation Criteria
+
+### Dev Preview -> Tech Preview
+
+The necessary feature for this solution (ValidatingAdmissionPolicy) is already available, even though it is in Tech Preview (it is in beta but disabled by default upstream). The evolution of this solution ties with the evolution of [ValidatingAdmissionPolicy](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md) and the decisions to graduate it downstream on Openshift.
+
+### Tech Preview -> GA
```

Comment:
documentation (would be better as a tool in `oc`) that can inspect an OCP cluster (or offline audit log) and produce a list of permissions that are needed to avoid disruption and with an option to auto-produce a directory of manifests that would apply the permissions.